### PR TITLE
Auxiliary array variable

### LINKED ISSUE #15811: Enable array auxiliary variables
## Reason
Array auxiliary variables can be useful.

## Design
Basic machinery for array variables are already in place. We just need to remove some `mooseError`s in the aux system and make use of it. 

## Impact
New capability.

### PR DESCRIPTION
**Pull request by YaqiWang**
Close #15811.

### CODE REVIEW

**Comment by YaqiWang** [comment]
This is done and ready for review.

**Comment by loganharbour** [comment]
Please push a change to this to forcefully re-trigger testing (full testing is now available).

You can do the following:
```
git commit --amend
```
and save the amend message that comes up without changing it (by typing `:wq` if in vim). This will forcefully change the commit hash of the last commit, which will trigger a change. Then forcefully push:
```
git push --force
```

**Comment by YaqiWang** [comment]
I give up the first commit. I have no idea why it is causing bunch of tests with ICs failing.

**Comment by YaqiWang** [comment]
This is ready for review.

**Comment by YaqiWang** [comment]
@lindsayad Want to take a look at this? You are an expert on templating which is this PR pretty much about.

**Comment by lindsayad** [comment]
I should be able to get to this today. I'll want the merge commit stripped out, but you can do that when I do a full review

**Comment by YaqiWang** [comment]
Thanks @lindsayad! That merge commit was generated when I practicing github merging conflict resolver, ;-) I can easily remove it and merge it into previous commits.

File: framework/include/auxkernels/AuxKernel.h [hunk 1]
```diff
@@ -48,6 +49,9 @@ InputParameters validParams<AuxKernel>();
 template <>
 InputParameters validParams<VectorAuxKernel>();
 
+template <>
+InputParameters validParams<ArrayAuxKernel>();
```
Comment (lindsayad) [inline framework/include/auxkernels/AuxKernel.h:53]: Legacy code. Can remove

File: framework/include/auxkernels/FunctionArrayAux.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "AuxKernel.h"
+
+// Forward Declarations
+class FunctionArrayAux;
```
Comment (lindsayad) [inline framework/include/auxkernels/FunctionArrayAux.h:15]: Can remove

File: framework/src/auxkernels/AuxKernel.C [hunk 3]
```diff
@@ -358,6 +364,75 @@ AuxKernelTempl<ComputeValueType>::compute()
...
+    if (_n_local_dofs == 1) /* p0 */
+    {
+      RealEigenVector value = RealEigenVector::Zero(_var.count());
+      for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+        value += _JxW[_qp] * _coord[_qp] * computeValue();
+      value /= (_bnd ? _current_side_volume : _current_elem_volume);
+      // update the variable data referenced by other kernels.
+      // Note that this will update the values at the quadrature points too
+      // (because this is an Elemental variable)
+      _var.setNodalValue(value);
```
Comment (lindsayad) [inline framework/src/auxkernels/AuxKernel.C:395]: Ugh, I don't like this, but this is what we do elsewhere in this object

File: framework/doc/content/source/auxkernels/FunctionArrayAux.md [hunk 4]
```diff
@@ -0,0 +1 @@
+!template load file=stubs/moose_object.md.template name=FunctionArrayAux syntax=/AuxKernels/FunctionArrayAux
```
Comment (lindsayad) [inline framework/doc/content/source/auxkernels/FunctionArrayAux.md:1]: Let's try not to add stubs. If this this is going to be a framework level object, let's add some documentation. It looks sloppy coming to an "under construction" page.

**Review by lindsayad — CHANGES_REQUESTED** [review]
Looks pretty straightforward and clean

File: framework/include/auxkernels/AuxKernel.h (hunk 1 above)
Comment (YaqiWang) [reply framework/include/auxkernels/AuxKernel.h:53]: We actually have `validParams<VectorAuxKernel>` and `validParams<AuxKernel>` right above this. So it is better to keep this and remove all three together in the future.

**Comment by YaqiWang** [comment]
@lindsayad comments addressed.

**Review by lindsayad — APPROVED** [review]
Thanks for the documentation!

**Pull request merged**